File: ia_assistant/java_backend/src/main/java/com/ia_assistant/os_core/memory/replacement/PageReplacementAlgorithm.java
package com.ia_assistant.os_core.memory.replacement;

/**
 * Interfaz para las estrategias de reemplazo de páginas.
 * Esto permite que el MemoryManager utilice diferentes algoritmos (FIFO, LRU, etc.) de forma intercambiable.
 */
public interface PageReplacementAlgorithm {

    /**
     * Se llama cuando se produce un fallo de página y no hay marcos libres.
     * El algoritmo debe decidir qué marco de página debe ser la "víctima".
     * @return El número del marco físico que ha sido seleccionado para reemplazo.
     */
    Integer findVictimFrame();

    /**
     * Notifica al algoritmo que una página ha sido cargada en un marco.
     * El algoritmo debe empezar a rastrear este marco.
     * @param frameNumber El número del marco que ahora está ocupado.
     */
    void onPageLoad(int frameNumber);

    /**
     * Notifica al algoritmo que se ha accedido a una página que ya estaba en memoria.
     * Esencial para algoritmos como LRU, que necesitan saber sobre los accesos.
     * @param frameNumber El número del marco al que se accedió.
     */
    void onPageAccess(int frameNumber);

    /**
     * Notifica al algoritmo que un marco ha sido liberado (ej. cuando un proceso termina).
     * El algoritmo debe dejar de rastrear este marco.
     * @param frameNumber El número del marco que ha sido liberado.
     */
    void onFrameFree(int frameNumber);
    
    /**
     * Obtiene el nombre de la estrategia actual (ej. "FIFO", "LRU").
     * @return El nombre del algoritmo.
     */
    String getName();

     /**
     * Devuelve una representación en String del estado interno del algoritmo (ej. la cola o lista).
     * @return Un String que describe el estado de la cola/lista de reemplazo.
     */
    String getStatus();
}

File: src/AuraOS/core/memory/PageReplacementAlgorithm.ts

// src/AuraOS/core/memory/PageReplacementAlgorithm.ts

/**
 * Define el contrato para los algoritmos de reemplazo de páginas.
 * Inspirado en la arquitectura de backend de Java.
 */
export interface PageReplacementAlgorithm {
  /**
   * Encuentra un marco "víctima" para reemplazar según la estrategia del algoritmo.
   * @returns El número del marco a ser reemplazado.
   */
  findVictimFrame(): number;

  /**
   * Notifica al algoritmo que una página nueva ha sido cargada en un marco.
   * @param frameNumber El número del marco donde se cargó la página.
   */
  onPageLoad(frameNumber: number): void;

  /**
   * Notifica al algoritmo que se ha accedido a una página que ya estaba en memoria.
   * Esencial para algoritmos como LRU.
   * @param frameNumber El número del marco al que se accedió.
   */
  onPageAccess(frameNumber: number): void;

  /**
   * Notifica al algoritmo que un marco ha sido liberado.
   * @param frameNumber El número del marco liberado.
   */
  onFrameFree(frameNumber: number): void;

  /**
   * Obtiene el nombre de la estrategia (ej. "FIFO", "LRU").
   * @returns El nombre del algoritmo.
   */
  getName(): string;

  /**
   * Devuelve una representación del estado interno del algoritmo (ej. la cola o lista).
   * @returns Un string que describe el estado.
   */
  getStatus(): string;
}
